alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/combiner/AbstractMessageCombiner.java

```diff
@@ -0,0 +1,156 @@
+        for (Set<ProviderMessageContent> groupedMessages : messagesGroupedByKey.values()) {
+            combineGroupedMessages(groupedMessages)
+                .ifPresent(combinedMessages::add);
+        }
+        return combinedMessages;
+    }
+
+    public ProviderMessageContent createNewMessage(ProviderMessageContent oldMessage, Collection<ComponentItem> componentItems) throws AlertException {
+        LinkableItem provider = oldMessage.getProvider();
+        LinkableItem topic = oldMessage.getTopic();
+        Optional<LinkableItem> optionalSubTopic = oldMessage.getSubTopic();
+        String subTopicName = optionalSubTopic.map(LinkableItem::getName).orElse(null);
+        String subTopicValue = optionalSubTopic.map(LinkableItem::getValue).orElse(null);
+        String subTopicUrl = optionalSubTopic.flatMap(LinkableItem::getUrl).orElse(null);
+        ItemOperation action = oldMessage.getAction().orElse(null);
+        Long notificationId = oldMessage.getNotificationId().orElse(null);
+
+        return new ProviderMessageContent.Builder()
+                   .applyProvider(provider.getValue(), provider.getUrl().orElse(null))
+                   .applyTopic(topic.getName(), topic.getValue(), topic.getUrl().orElse(null))
+                   .applySubTopic(subTopicName, subTopicValue, subTopicUrl)
+                   .applyAction(action)
+                   .applyNotificationId(notificationId)
+                   .applyAllComponentItems(componentItems)
+                   .build();
+    }
+
+    protected LinkedHashSet<ComponentItem> gatherComponentItems(Collection<ProviderMessageContent> groupedMessages) {
+        List<ComponentItem> allComponentItems = groupedMessages
+                                                    .stream()
+                                                    .map(ProviderMessageContent::getComponentItems)
+                                                    .flatMap(Set::stream)
+                                                    .collect(Collectors.toList());
+        return combineComponentItems(allComponentItems);
+    }
+
+    protected LinkedHashSet<ComponentItem> combineComponentItems(List<ComponentItem> allComponentItems) {
+        // The amount of collapsing we do makes this impossible to map back to a single notification.
+        Map<String, ComponentItem> keyToItems = new LinkedHashMap<>();
+        for (ComponentItem componentItem : allComponentItems) {
+            Set<LinkableItem> componentAttributes = componentItem.getComponentAttributes();
+            String key = componentItem.createKey(true, true);
+            ComponentItem oldItem = keyToItems.get(key);
+
+            Set<LinkableItem> combinedAttributes = new LinkedHashSet<>();
+            if (null != oldItem) {
+                combinedAttributes.addAll(oldItem.getComponentAttributes());
+                combinedAttributes.addAll(componentAttributes);
+            } else {
+                combinedAttributes.addAll(componentAttributes);
```

Comment:
this can be simplified. Make the null check and add the oldItems if not null. Then always add the rest of componentAttributes.